paddle/optimizer/adadelta_optimizer.cc

```diff
@@ -0,0 +1,63 @@
+#include "adadelta_optimizer.h"
+#include <algorithm>
+#include <cmath>
+
+namespace paddle {
+namespace optimizer {
+
+void AdadeltaOptimizer::Update(const Tensor* gradient) {
+  num_sample_passed_ += 1;
+  double learning_rate = lr_policy_->LearningRate(num_sample_passed_);
+  Tensor& param = *parameter_;
+  const Tensor& grad = *gradient;
+  Tensor& accum_g = *accum_gradient_;
+  Tensor& accum_d = *accum_delta_;
+  Tensor& update_d = *update_delta_;
+  for (size_t i = 0; i < param.size(); ++i) {
+    accum_g[i] = rho_ * accum_g[i] + (1.0 - rho_) * grad[i] * grad[i];
+
+    update_d[i] = std::sqrt(accum_d[i] + epsilon_) /
+                  std::sqrt(accum_g[i] + epsilon_) * grad[i];
+
+    accum_d[i] = rho_ * accum_d[i] + (1.0 - rho_) * update_d[i] * update_d[i];
+
+    param[i] -= learning_rate * update_d[i] + learning_rate * decay_ * param[i];
+  }
+}
+
+const char* AdadeltaOptimizer::SerializeState(int* state_len) {
+  OptimizerState state;
+  state.set_learning_rate(lr_policy_->LearningRate(num_sample_passed_));
+  state.set_num_sample_passed(num_sample_passed_);
+
+  TensorToProto(*parameter_, state.mutable_parameter());
+  TensorToProto(*accum_gradient_, state.mutable_accum_gradient());
+  TensorToProto(*accum_delta_, state.mutable_accum_delta());
+  TensorToProto(*update_delta_, state.mutable_update_delta());
+  state.set_nesterov(epsilon_);
+  state.set_momentum(rho_);
+  state.set_decay(decay_);
+  // can be used when memory alignment to system
+  *state_len += CalStateSize(parameter_,
+                             accum_gradient_,
+                             accum_delta_,
+                             update_delta_,
+                             rho_,
+                             epsilon_,
+                             decay_);
+  return state.SerializeAsString().c_str();
+}
+
+void AdadeltaOptimizer::DeSerializeState(const std::string& str) {
+  OptimizerState state;
+  state.ParseFromString(str);
+  lr_policy_->set(state.learning_rate());
+  num_sample_passed_ = state.num_sample_passed();
```

Comment:
firstly, `epsilon_`, `rho_`, `decay_` is not missing by careless. Consider that we have to call `create_with_config` function signature with config, which contains these three parameters. And these ones is constant varible during training process. Do we really need to save them since we already have to save the config proto file? 
If we want the training state self-contained, these constant hyperparameters should be there.

secondly, In that format, obviously it is easy to reading. But need a pair of `config/state` for every optimizer, which lead to  copy/paste config config and Serialize/Deserialize code many times, we should find another elegant and idiomatic way to do that. The idea is the same way goes with https://github.com/caffe2/caffe2/blob/master/caffe2/core/qtensor_serialization.h